File: engine/yaml_helpers.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Shared YAML loading helpers for all engine scripts.
# Consolidates safe_load patterns from contract.rb, inbox.rb, and validate.rb.

require "psych"
require "time"
require "date"

module YamlHelpers
  PERMITTED_CLASSES = [Time, Date].freeze

  # Load a YAML file safely. Returns the expected type's empty value if file
  # is missing or contains nil. Raises on type mismatch.
  def self.safe_load(path, expected_class)
    return(expected_class == Hash ? {} : []) unless File.exist?(path)
    content = File.read(path)
    data = Psych.safe_load(content, permitted_classes: PERMITTED_CLASSES, aliases: false)
    return(expected_class == Hash ? {} : []) if data.nil?
    raise "YAML document has unexpected type in #{path}" unless data.is_a?(expected_class)
    data
  end

  # Load and normalize Time/Date scalars to ISO 8601 strings.
  def self.safe_load_normalized(path, expected_class)
    data = safe_load(path, expected_class)
    normalize_scalar_values(data)
  end

  def self.normalize_scalar_values(value)
    case value
    when Time then value.utc.iso8601
    when Date then value.iso8601
    when Array then value.map { |v| normalize_scalar_values(v) }
    when Hash then value.transform_values { |v| normalize_scalar_values(v) }
    else value
    end
  end
end

File: src/superharness/engine/yaml_helpers.py
"""Mirrors YamlHelpers Ruby module.

Shared YAML loading helpers for all engine scripts.
Consolidates safe_load patterns from contract.py, inbox.py, and validate.py.
"""
from __future__ import annotations

import io
import os
from datetime import date, datetime
from typing import Union

import yaml  # fallback (PyYAML)

try:
    from ruamel.yaml import YAML as RuamelYAML

    _ruamel_rt = RuamelYAML()  # round-trip (default type)
    _ruamel_rt.preserve_quotes = True
except ImportError:
    _ruamel_rt = None  # type: ignore[assignment]


def safe_load(path: str, expected_class: type) -> Union[dict, list]:
    """Load a YAML file safely.

    Returns the expected type's empty value if the file is missing or contains
    nil. Raises TypeError on type mismatch.
    """
    empty: Union[dict, list] = {} if expected_class is dict else []
    if not os.path.exists(path):
        return empty
    with open(path, "r") as f:
        content = f.read()
    data = yaml.safe_load(content)
    if data is None:
        return empty
    if not isinstance(data, expected_class):
        raise TypeError(
            f"YAML document has unexpected type in {path}: "
            f"expected {expected_class.__name__}, got {type(data).__name__}"
        )
    return data


def _normalize(value: object) -> object:
    """Normalize Time/Date scalars to ISO 8601 strings, like Ruby yaml_helpers."""
    if isinstance(value, datetime):
        return value.isoformat()
    if isinstance(value, date):
        return value.isoformat()
    if isinstance(value, list):
        return [_normalize(v) for v in value]
    if isinstance(value, dict):
        return {k: _normalize(v) for k, v in value.items()}
    return value


def safe_load_normalized(path: str, expected_class: type) -> Union[dict, list]:
    """Load and normalize Time/Date scalars to ISO 8601 strings."""
    data = safe_load(path, expected_class)
    result = _normalize(data)
    # _normalize preserves the dict/list type so cast is safe
    return result  # type: ignore[return-value]


def round_trip_load(path: str) -> object:
    """Load YAML preserving comments and formatting."""
    if _ruamel_rt is None:
        return safe_load(path, dict)
    with open(path, "r") as f:
        return _ruamel_rt.load(f)


def round_trip_dump(data: object, path: str) -> None:
    """Write YAML preserving comments and formatting."""
    if _ruamel_rt is None:
        with open(path, "w") as f:
            yaml.dump(data, f)
        return
    buf = io.StringIO()
    _ruamel_rt.dump(data, buf)
    with open(path, "w") as f:
        f.write(buf.getvalue())
